File: Haskell/Cliente.hs
module Cliente where
import Mensagens
import Utils
import System.IO
import Mensagens

menuCliente:: IO()
menuCliente  = do
    putStr("\n----------Menu do Cliente----------\n")
    putStr("1 - Listar veículos disponíves\n")
    putStr("2 - Escolher veículo\n")
    putStr("3 - Pesquisar veículos\n")
    putStr("4 - Voltar para o menu principal\n")
    putStr("Opção: ")
    op <- readLn:: IO Int
    chamadaCliente op


chamadaCliente:: Int -> IO()
chamadaCliente op 
    | op == 1 = do 
        veiculosDisponiveis
        menuCliente
    | op == 2 = do 
        escolherVeiculoCliente
        menuCliente
    | op == 3 = do 
        menuPesquisaVeiculos
        menuCliente
    | op == 4 = putStr ""
    | otherwise = do
        putStr("Opção inválida, digite novamente\n")
        menuCliente

escolherVeiculoCliente:: IO()
escolherVeiculoCliente = do
    veiculosDisponiveis
    clientesCadastrados

    putStr "\nInforme seu código: "
    idCliente <- readLn:: IO Int

    putStr "Informe o código do carro: "
    idCarro <- readLn:: IO Int

    putStr "Tipo de contrato:\n"
    putStr "1 - Diário\n"
    putStr "2 - Mensal\n"
    tipoContrato <- readLn :: IO Int

    putStr "Quantidade de dias/meses: "
    tempo <- readLn:: IO Int

    arqClientes <- openFile "arquivos/clientes.txt" ReadMode
    contents <- hGetContents arqClientes
    let clientes = lines contents

    if Utils.procuraCliente clientes idCliente == [] then do 
        putStr Mensagens.clienteNaoEncontrado
        hClose arqClientes
    else do
        if clientes == clientes then do
            hClose arqClientes

            arqCarros <- openFile "arquivos/carros.txt" ReadMode
            contents2 <- hGetContents arqCarros
            let carros = lines contents2

            if Utils.procuraCarro carros idCarro == [] then do
                putStr Mensagens.carroNaoEncontrado
                hClose arqCarros
            else do
                if carros == carros then do
                    hClose arqCarros

                    Utils.escolherCarroCliente clientes idCliente carros idCarro (if tipoContrato == 1 then "diario" else "mensal") tempo
                else putStr ""
        
        else putStr ""

menuPesquisaVeiculos:: IO()
menuPesquisaVeiculos = do
    putStr "Tipo de pesquisa:\n"
    putStr "1 - Modelo\n"
    putStr "2 - Ano\n"
    putStr "3 - Cor\n"
    op <- readLn:: IO Int

    opcaoPesquisa op

opcaoPesquisa:: Int -> IO()
opcaoPesquisa op = do
    arqCarros <- openFile "arquivos/carros.txt" ReadMode
    contents2 <- hGetContents arqCarros
    let carros = lines contents2

    if op > 3 || op < 1 then putStr "Opção inválida\n"
    else do
        if op == 1 then putStr "Modelo: "
        else if op == 2 then putStr "Ano: "
        else putStr "Cor: "
        element <- getLine
        Utils.escreveCarrosElem carros element op
        hClose arqCarros

clientesCadastrados:: IO()
clientesCadastrados = do
    putStrLn "\nClientes cadastrados:"
    
    arq <- readFile "arquivos/clientes.txt"
    let linhas = lines arq
    Utils.escreveClientes linhas

veiculosDisponiveis:: IO()
veiculosDisponiveis = do
    putStrLn "\nVeículos disponíveis:"

    arq <- readFile "arquivos/carros.txt"
    let linhas = lines arq
    Utils.escreveCarrosDisponiveis linhas

File: Prolog/cliente.pl
:- include('mensagens.pl').
:- include('utils.pl').

menuCliente:-
    writeln('\n----------Menu do Cliente----------\n'),
    write('1 - Listar veiculos disponives\n'),
    write('2 - Escolher veiculo\n'),
    write('3 - Pesquisar veiculos\n'),
    write('4 - Voltar para o menu principal\n'),
    write('Opcao: '),
    read(Opcao),
    chamadaCliente(Opcao).

chamadaCliente(1):-
    veiculosDisponiveis,
    menuCliente.
chamadaCliente(2):-
    escolherVeiculoCliente,
    menuCliente.
chamadaCliente(3):-
    menuPesquisaVeiculos, 
    menuCliente.
chamadaCliente(4):- write('').
chamadaCliente(_):-
    opcaoInvalida, 
    menuCliente.

veiculosDisponiveis:- 
    writeln('\nVeiculos disponiveis:'),
    exibirCarrosDisp('arquivos/carros.csv').

escolherVeiculoCliente:-
    clientesCadastrados,
    write('\nInforme seu codigo: '),
    read(IdCliente),
    existeCliente(IdCliente, 'arquivos/clientes.csv', R),
    (R == true -> veiculosDisponiveis,
    write('\nInforme o codigo do carro: '),
    read(IdCarro),

    writeln('Tipo de contrato:'),
    writeln('1 - Diario'),
    writeln('2 - Mensal'),
    read(Tipo),
    write('Quantidade de dias/meses: '),
    read(Tempo),
    
    (Tipo =:= 1 -> escolherCarroCliente(IdCliente, IdCarro, 'diario', Tempo); escolherCarroCliente(IdCliente, IdCarro, 'mensal', Tempo)),
    writeln('Carro alugado com sucesso!');
    writeln('Cliente nao existe!')).

menuPesquisaVeiculos:- 
    writeln('\nTipo de pesquisa:'),
    writeln('1 - Modelo'),
    writeln('2 - Ano'),
    writeln('3 - Cor'),
    write('Opcao: '),
    read(Op),
    opcaoPesquisa(Op).

opcaoPesquisa(1):-
    writeln('\nModelo: '),
    read(Modelo),
    pesquisaCarroModelo(Modelo, 'arquivos/carros.csv').
opcaoPesquisa(2):-
    writeln('\nAno: '),
    read(Ano),
    pesquisaCarroAno(Ano, 'arquivos/carros.csv').
opcaoPesquisa(3):-
    writeln('\nCor: '),
    read(Cor),
    pesquisaCarroCor(Cor, 'arquivos/carros.csv').
opcaoPesquisa(_):-
    opcaoInvalida,
    menuPesquisaVeiculos.

